### PR TITLE
Add scale-only version of the HQQ algorithm for IntxWeightOnlyConfig/Int8DynamicActivationIntxWeightConfig

### PR DESCRIPTION
**Pull request by metascroy**
This PR introduces a scale-only version of the HQQ algorithm for  IntxWeightOnlyConfig/Int8DynamicActivationIntxWeightConfig, which we find improves model quality on mmlu from 0.50 to 0.55 when quantizing Gemma3-4B.

### CODE REVIEW

**Comment by jerryzh168** [comment]
Thanks! Add some unit tests for this path as well?

File: torchao/quantization/quant_api.py [hunk 1]
```diff
@@ -2202,6 +2216,7 @@ def __post_init__(self):
         assert self.mapping_type in [
             MappingType.ASYMMETRIC,
             MappingType.SYMMETRIC,
+            MappingType.SYMMETRIC_NO_CLIPPING_ERR,
```
Comment (mergennachin) [inline torchao/quantization/quant_api.py:2219]: need to update the assert error message

**Comment by metascroy** [comment]
> Thanks! Add some unit tests for this path as well?

Added new E2E test

File: torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    """
+    Uses `torchao.quantization.quant_primitives.choose_qparams_affine`
+    """
+    AFFINE = "affine"
+
+    """
+    Uses `torchao.quantization.quant_primitives._choose_qparams_and_quantize_scale_only_hqq`
+    """
+    HQQ = "hqq"
```
Comment (mergennachin) [inline torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py]: There are two variants of HQQ today, which makes it confusing:

_choose_qparams_and_quantize_affine_hqq -- which is used in version 1 of the config when use_hqq=True

_choose_qparams_and_quantize_scale_only_hqq - this one

File: torchao/quantization/quant_primitives.py [hunk 3]
```diff
@@ -2125,6 +2126,99 @@ def _choose_qparams_and_quantize_affine_hqq(
...
+    for _ in range(max(1, iters)):
+        # Quantize using current scale
+        # Qg.shape = [n, n_groups, group_size]
+        Qg = _r(Wg / scale.unsqueeze(-1)).clamp(qmin, qmax)
+
+        # Solve least-square problem min_{s} ||Wg - s * Qg||^2 and project
+        # solution onto positive space, or take previous value
+        num = (Wg * Qg).sum(dim=2, dtype=torch.float32)  # [n, n_groups]
+        den = (Qg * Qg).sum(dim=2, dtype=torch.float32)  # [n, n_groups]
+        scale = torch.where(den > 0, num / den, prev_scale)
```
Comment (mergennachin) [inline torchao/quantization/quant_primitives.py:2200]: can `den` be really really small? dividing num by den may result really big scales

File: torchao/quantization/qat/api.py [hunk 4]
```diff
@@ -270,6 +256,12 @@ def _qat_config_transform(
                 f"Encountered unexpected module {module}, should never happen"
             )
         if base_config is not None:
+            # If passing custom scales and zero points, we need to disable the choose_qparam_algorithm on the config
+            if has_custom_scale_and_zero_point and hasattr(
+                base_config, "intx_choose_qparams_algorithm"
+            ):
+                base_config = copy.deepcopy(base_config)
+                base_config.intx_choose_qparams_algorithm = None
```
Comment (mergennachin) [inline torchao/quantization/qat/api.py:266]: Can you print a debug log that it is disabled?

File: torchao/quantization/qat/api.py [hunk 5]
```diff
@@ -270,6 +256,12 @@ def _qat_config_transform(
                 f"Encountered unexpected module {module}, should never happen"
             )
         if base_config is not None:
+            # If passing custom scales and zero points, we need to disable the choose_qparam_algorithm on the config
+            if has_custom_scale_and_zero_point and hasattr(
+                base_config, "intx_choose_qparams_algorithm"
+            ):
```
Comment (mergennachin) [inline torchao/quantization/qat/api.py:263]: do you wanna specifically check for Int8DynamicActivationIntxWeightConfig and IntxWeightOnlyConfig types only?

File: torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py (hunk 2 above)
Comment (scottroy) [reply torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py]: These configs are used in ExecuTorch, and never supported HQQ in any previous version.

use_hqq was used in version 1 of the int4 server config.

If you think it would be clearer, I could call the algorithm enum HQQ_SCALE_ONLY or HQQ_NO_ZERO_POINT

File: torchao/quantization/quant_primitives.py (hunk 3 above)
Comment (scottroy) [reply torchao/quantization/quant_primitives.py:2200]: den is a sum of square integers, so it will be at least 1 if positive.

File: torchao/quantization/qat/api.py (hunk 5 above)
Comment (scottroy) [reply torchao/quantization/qat/api.py:263]: I originally did this, but it causes a circular import error because the file that defines the configs also imports from this file.

File: torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py (hunk 2 above)
Comment (mergennachin) [reply torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py]: > If you think it would be clearer, I could call the algorithm enum HQQ_SCALE_ONLY or HQQ_NO_ZERO_POINT

Yeah, let's do HQQ_SCALE_ONLY then

**Comment by mergennachin** [comment]
This is great btw

Why would one ever not do HQQ_scale_only on XNNPACK on ET if it improves the accuracy compared to naive? It doesn't take a long time to quantize, right? Roughly a few minutes?

Hope we can make it to the 0.14 branch cut, which is Oct 6.

File: test/quantization/quantize_/workflows/intx/test_intx_unpacked_to_int8_tensor.py [hunk 6]
```diff
@@ -61,6 +61,22 @@ def test_linear(self):
         error = compute_error(original, quantized)
         self.assertTrue(error > 20)
 
+    def test_hqq(self):
+        dtype = torch.bfloat16
+        device = "cpu"
+        config = IntxWeightOnlyConfig(
```
Comment (mergennachin) [inline test/quantization/quantize_/workflows/intx/test_intx_unpacked_to_int8_tensor.py:67]: Also another test for Int8DynamicActivationIntxWeightConfig too

File: torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+from enum import Enum
+
+
+# can switch to StrEnum (https://docs.python.org/3/library/enum.html#enum.StrEnum)
+# after python 3.10 is end of life (https://devguide.python.org/versions/)
+class IntxChooseQParamsAlgorithm(str, Enum):
```
Comment (mergennachin) [inline torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py:12]: Also, do we need to introduce yet another class? 

Can we just extend existing Int4ChooseQParamsAlgorithm and add affine and hqq_scale_only? 

And then rename/promote Int4ChooseQParamsAlgorithm to IntxChooseQParamsAlgorithm in a follow-up PR?

**Comment by metascroy** [comment]
> This is great btw
> 
> Why would one ever not do HQQ_scale_only on XNNPACK on ET if it improves the accuracy compared to naive? It doesn't take a long time to quantize, right? Roughly a few minutes?
> 
> Hope we can make it to the 0.14 branch cut, which is Oct 6.

Yes, I want to measure on some more models and benchmarks and if results are neutral to positive, I'll update the default in etLLM.

File: torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py (hunk 7 above)
Comment (metascroy) [reply torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py:12]: In torchao's refactor (removing AffineQuantizedTensor), the direction is for subclasses to not share higher-level abstractions, but instead define their own enums.  This is how packing format works as well (intx_packing_format for the intx subclass, and int4_packing_format for the int4 subclass).

I'll let @jerryzh168 comment here as well

Comment (mergennachin) [reply torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py:12]: Will defer to @jerryzh168 then

Comment (jerryzh168) [reply torchao/quantization/quantize_/workflows/intx/intx_choose_qparams_algorithm.py:12]: yeah we want local abstractions instead of global abstractions unless it's required.

File: torchao/quantization/qat/api.py
Comment (jerryzh168) [inline torchao/quantization/qat/api.py:1]: btw, are the changes in this file tested as well?

Comment (metascroy) [reply torchao/quantization/qat/api.py:1]: I think this exisiting unit test covers them: https://github.com/pytorch/ao/blob/main/test/quantization/test_qat.py#L2321

File: torchao/quantization/qat/api.py (hunk 4 above)
Comment (jerryzh168) [reply torchao/quantization/qat/api.py:266]: oh so the new algorithm is not supported, should you error out instead when user specifies the hqq algorithm instead of silent ignore?

Comment (scottroy) [reply torchao/quantization/qat/api.py:266]: This is not silent ignore.  This is used in QAT range learning, where the scale/zero_point are passed when constructing the subclass.  But I changed the subclass to require you specify choose algorithm as None if you pass in scale/zero.
